dangerzone/updater/signatures.py

```diff
@@ -0,0 +1,597 @@
+import json
+import platform
+import re
+import subprocess
+import tarfile
+from base64 import b64decode, b64encode
+from dataclasses import dataclass
+from functools import reduce
+from hashlib import sha256
+from io import BytesIO
+from pathlib import Path, PurePath
+from tempfile import NamedTemporaryFile, TemporaryDirectory
+from typing import Callable, Dict, List, Optional, Tuple
+
+from .. import container_utils as runtime
+from .. import errors as dzerrors
+from ..util import get_resource_path
+from . import cosign, errors, log, registry
+
+try:
+    import platformdirs
+except ImportError:
+    import appdirs as platformdirs  # type: ignore[no-redef]
+
+
+def appdata_dir() -> Path:
+    return Path(platformdirs.user_data_dir("dangerzone"))
+
+
+# RELEASE: Bump this value to the log index of the latest signature
+# to ensure the software can't upgrade to container images that predates it.
+DEFAULT_LOG_INDEX = 0
+
+DEFAULT_PUBKEY_LOCATION = get_resource_path("freedomofpress-dangerzone-pub.key")
+SIGNATURES_PATH = appdata_dir() / "signatures"
+LAST_LOG_INDEX = SIGNATURES_PATH / "last_log_index"
+DANGERZONE_MANIFEST = "dangerzone.json"
+
+__all__ = [
+    "verify_signature",
+    "load_and_verify_signatures",
+    "store_signatures",
+    "verify_offline_image_signature",
+]
+
+
+def signature_to_bundle(sig: Dict) -> Dict:
+    """Convert a cosign-download signature to the format expected by cosign bundle."""
+    bundle = sig["Bundle"]
+    payload = bundle["Payload"]
+    return {
+        "base64Signature": sig["Base64Signature"],
+        "Payload": sig["Payload"],
+        "cert": sig["Cert"],
+        "chain": sig["Chain"],
+        "rekorBundle": {
+            "SignedEntryTimestamp": bundle["SignedEntryTimestamp"],
+            "Payload": {
+                "body": payload["body"],
+                "integratedTime": payload["integratedTime"],
+                "logIndex": payload["logIndex"],
+                "logID": payload["logID"],
+            },
+        },
+        "RFC3161Timestamp": sig["RFC3161Timestamp"],
+    }
+
+
+def verify_signature(signature: dict, image_digest: str, pubkey: Path) -> None:
+    """
+    Verifies that:
+
+    - the signature has been signed by the given public key
```

Comment:
```suggestion
    - the signature matches the given public key
```

Reply:
I went with a one liner in the end, see 21cbbb6760c60613f2877d3ae9d3b4ce7174c08c